File: java/leetcode/src/Task94.java
import org.junit.jupiter.api.Test;

import java.util.*;

public class Task94 {

    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();

        recursion(root,res);

        return res;
    }

    void recursion(TreeNode root,List<Integer> res){
        if(root == null){
            return;
        }

        recursion(root.left,res);
        res.add(root.val);
        recursion(root.right,res);
    }

    //iteration
    public List<Integer> inorderTraversal2(TreeNode root) {
        List<Integer> res = new ArrayList<>();

        if(root == null){
            return res;
        }
        Stack<TreeNode> stack = new Stack<>();

        TreeNode cur = root;
        while (cur != null ||!stack.isEmpty()){
            while (cur != null){
                stack.push(cur);
                cur = cur.left;
            }
            cur = stack.pop();
            res.add(cur.val);
            cur = cur.right;
        }

        return res;
    }

    @Test
    void test(){
        TreeNode root = new TreeNode(1);
        root.right(2).right.left(3);
        System.out.println(inorderTraversal2(root));
    }
}

File: py/Task94.py
# Definition for a binary tree node.
from typing import List


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    # dfs
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        res = []

        def dfs(root: TreeNode):
            if not root:
                return
            dfs(root.left)
            res.append(root.val)
            dfs(root.right)
        dfs(root)
        return res

    # using stack
    def inorderTraversal2(self, root: TreeNode) -> List[int]:
        res = []
        stack = []
        cur = root
        while cur or stack:
            if cur:
                stack.append(cur)
                cur = cur.left
            else:
                cur = stack.pop()
                res.append(cur.val)
                cur = cur.right
        return res
